gt/lua_lpm.c

```diff
@@ -0,0 +1,395 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <arpa/inet.h>
+#include <lauxlib.h>
+
+#include <rte_lcore.h>
+
+#include "lua_lpm.h"
+#include "luajit-ffi-cdata.h"
+#include "gatekeeper_net.h"
+#include "gatekeeper_lpm.h"
+#include "gatekeeper_fib.h"
+
+static int
+l_str_to_prefix(lua_State *l)
+{
+	int ret;
+	struct ipaddr ip_addr;
+
+	/* First argument must be an IP prefix string. */
+	const char *prefix_str = luaL_checkstring(l, 1);
+
+	if (lua_gettop(l) != 1)
+		luaL_error(l, "Expected one argument, however it got %d arguments",
+			lua_gettop(l));
+
+	ret = parse_ip_prefix(prefix_str, &ip_addr);
+	if (ret < 0 || ip_addr.proto == ETHER_TYPE_IPv6)
+		luaL_error(l, "gk: failed to parse an IPv4 prefix!");
+
+	lua_pushinteger(l, ip_addr.ip.v4.s_addr);
+	lua_pushinteger(l, ret);
+
+	return 2;
+}
+
+#define CTYPE_STRUCT_IN6_ADDR_REF "struct in6_addr &"
+
+static int
+l_str_to_prefix6(lua_State *l)
+{
+	int ret;
+	struct ipaddr ip_addr;
+	void *cdata;
+	uint32_t correct_ctypeid_in6_addr;
+
+	/* First argument must be an IP prefix string. */
+	const char *prefix_str = luaL_checkstring(l, 1);
+
+	if (lua_gettop(l) != 1)
+		luaL_error(l, "Expected one argument, however it got %d arguments",
+			lua_gettop(l));
+
+	ret = parse_ip_prefix(prefix_str, &ip_addr);
+	if (ret < 0 || ip_addr.proto == ETHER_TYPE_IPv4)
+		luaL_error(l, "gk: failed to parse an IPv6 prefix!");
+
+	correct_ctypeid_in6_addr = luaL_get_ctypeid(l,
+		CTYPE_STRUCT_IN6_ADDR_REF);
+	cdata = luaL_pushcdata(l, correct_ctypeid_in6_addr,
+		sizeof(struct in6_addr));
+	*(struct in6_addr *)cdata = ip_addr.ip.v6;
+
+	lua_pushinteger(l, ret);
+
+	return 2;
+}
+
+static int
+l_new_lpm(lua_State *l)
+{
+	struct rte_lpm_config lpm_conf;
+	struct rte_lpm **lpm;
```

Comment:
Style: rename `lpm` to `p_lpm` to emphasize that it's a pointer to pointer. Do the analog in `l_new_lpm6()`.

The other functions don't really need the type `struct rte_lpm **`, they will be happy with `struct rte_lpm *`. You just dereference the pointer before the assignment.